### ISSUE TITLE
Better scheme than changes.h

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#3154]
We keep running into the issue that proposed patches have to be rebased to fix the merge conflict on `changes.h`. This issue isn't going to go away as we seem to get more patches over time.

I'd like to discuss better ways to do this. My preferred approach is to have a directory `doc/news/changes/` into which patch authors simply put a file that corresponds to individual changelog entries. A script would then combine this into the `changes.h` file we send to doxygen. If we adopt the convention that these files start with `YYYY-MM-DD-...`, but are otherwise arbitrary, then we can combine them in a way that preserves the temporal order of the changelog, and would avoid all but the most unlikely merge conflicts.

Other suggestions that have been made over time include special comments in commit messages. I don't particularly like this because it's hard to auto-generate a `changes.h` file from that. It's also really difficult to deal with if someone accidentally gets the format wrong -- you can't fix it up after the fact.

Thoughts?

**Comment by davydden** [comment]
agree. I think the first approach (separate files `YYYY-MM-DD-...`) is more robust and easier to handle. One would probably also want to encode in filename the category where the particular changes go.

**Comment by bangerth** [comment]
Yes, or have subdirectories.

**Comment by davydden** [comment]
i think subdirectories are better.

**Comment by masterleinad** [comment]
I like this idea.

**Comment by tjhei** [comment]
> Thoughts?

Yes, this is a major pain point.

> A script would then combine this into the changes.h file we send to doxygen.

So this script would be run automatically when doing "make documentation" on the server and replace changes.h? We could then commit the result and delete the separate files occasionally and before a release.

> YYYY-MM-DD-..., but are otherwise arbitrary,

If you do YYY-MM-DD-githubusername (and for people like @bangerth with more than one patch per day, you can add a number :-) ).

Some more ideas:
- It would be great if these files above don't need the "<li>" or "*" indentation but just
```
New: bla.<br>
(Name, date)
```

While we are at it, we should talk about the three sections "major", "minor", and "incompatibilities". I never know the difference between the two former. We could get rid of them altogether, because "Changed:" makes it obvious that this is a breaking change.

**Comment by bangerth** [comment]
Yes, I'm sure we can write the script in a way that allows for that. Maybe just
```
  New: this that and the other:
  Author: my name
```
That shouldn't be terribly hard to parse.

As for the categories, I always thought of "major" (which is actually "general changes") as something that affects more than one function or class, or that has to do with tutorials, or that has to do with infrastructure such as the build system. Other suggestions welcome.

**Comment by davydden** [comment]
> As for the categories, I always thought of "major" (which is actually "general changes") as something that affects more than one function or class, or that has to do with tutorials, or that has to do with infrastructure such as the build system.

if we stick with that, i think this should be documented somewhere.

**Comment by bangerth** [comment]
I was thinking of putting a README file into `doc/news/changes/` that explains the format of these files and of file names.

**Comment by kronbichler** [comment]
I also like the idea of avoiding all those merge conflicts by separate files.

> As for the categories, I always thought of "major" (which is actually "general changes") as something that affects more than one function or class.

Among those software packages where I like reading the release notes, the "major" changes are something more substantial than just adding another variant of a function, rather something to get "excited" about (and a list that is sufficiently compact). We have the release papers to report about the most significant changes, so why not take something along those lines? Obviously, such a selection is something subjective, but the current scheme of putting 42 "general" changes and 77 "specific" is probably not the best way to write notes. In particular since many of the 42 items in the former category are not really groundbreaking.

My personal preference would be to have four lists: incompatibilities (as now), changes in the build system, major changes including new tutorials, and specific changes. To get into the "major" category, several developers should express their support. (For the current release, I only see step-55, step-56, and Manifold as clear candidates for the major list, maybe one or two more items depending on how selective we want to be.) But I might be too ambitious here...

**Comment by tjhei** [comment]
> To get into the "major" category,

Whenever we prepare a release, we go through the complete changelog and pick the ~10 most important things anyways (to list in the release text, email, etc.). We might as well decide at that point to move stuff into "major".

**Comment by tamiko** [comment]
Alternatively we could require merges of pull request to have dates / authors and a description.
More specifically, the developer that merges has to write up a small changelog entry (for example copied from the description of the pull request).

As soon we do a release we then simply combine the description of all merge commits into `changes.h`.

I agree with @tjhei that selecting major changes, important incompatibilities, etc. should be done at that point. With the current scheme we have to reorganize and clean up anyway.

This approach would have the advantage to
 - solve the merge conflict problem
 - be easy for new contributors (we don't have to explain the `changes.h` or another change-file business at all)
 - no additional complexity
 - a one-to-one correspondence between change and merge of a feature branch

**Comment by davydden** [comment]
> Alternatively we could require merges of pull request to have dates / authors and a description.
More specifically, the developer that merges has to write up a small changelog entry (for example copied from the description of the pull request).

Where does developer write this? If it is within GitHub, then it will be difficult to parse this info when we do a release. 

If description is to be written in a file, then, as i understand it, the only difference between this and the approach above is that the developer is to create a description file (in a separate PR?).

Either way i think it is much too complicated compared to the one file scheme with simple naming. I am in favor of the latter.

**Comment by davydden** [comment]
> My personal preference would be to have four lists: incompatibilities (as now), changes in the build system, major changes including new tutorials, and specific changes. 

I like this naming. I also agree with @tjhei the we don't need to be too strict about categories upfront. We can agree on those ~10 most important changes when preparing a release and move the rest into 'minor'/'specific' category.

**Comment by kronbichler** [comment]
I agree with @davydden and @tjhei that selecting the important changes when doing the release is the right thing to do since we go through the changes anyway.

**Comment by bangerth** [comment]
So do I hear a vote to have only two categories ("incompatible changes", "other changes" -- better name for second welcome) and that we would reclassify the latter at release time? That would make sense to me. I'm in favor.

I'm against the idea of having the reviewer write the entry.

**Comment by tamiko** [comment]
> I'm against the idea of having the reviewer write the entry.

Then make it a mandatory requirement that a pull request description is already a changes.h entry.

**Comment by tamiko** [comment]
@davydden 
> Where does developer write this? If it is within GitHub, then it will be difficult to parse this info when we do a release. 

In the description of the pull request. It would be copied to the actual merge commit by the commiter.
That way we would already have two pair of eyes going over the changes entry. The person doing the pull request, and the one doing the actual merge.

Later, we would simply have to parse all merge commits of master. This doesn't require github at all.

Such a scheme would be very much in line how a number of big projects handle their merges (and subsequent generation of a changelog).

**Comment by tamiko** [comment]
More specifically, I suggest the following workflow:
 - The author of the pull request provides a suitable changes entry, either in the description of the original pull request, or in a later comment.
 - The developer that merges the pull request clicks the "Merge pull request" button and copy and pastes the changes entry into the description field:
```
[A meaningful changes.h entry]

Authors: ..., ..., ...
Date: ...
```
With that we end up with a merge commit that reads
```
Merge: ... ...
Author: deal.II Developer
Date:   ...

    Merge pull request #xxxx from xxxx/xxxx
    
    [A meaningful changes.h entry]

    Authors: ..., ..., ...
    Date: ...
```

This would require a bit of discipline from developers. The great advantage would be that our beloved changes file is a mere
```
$ git log --first-parent
```
away. And we don't have to bother with files and a random file format and naming at all.

**Comment by tamiko** [comment]
For comparison, currently the output of `$git log --first-parent` looks like:
```
commit 95b09cd611d4116d18b77c9df6c22679c2ae169f
Merge: 0499c54 d94a9d9
Author: David Wells <drwells@vt.edu>
Date:   Sat Sep 24 00:03:54 2016 -0400

    Merge pull request #3156 from bangerth/fix-typo
    
    Fix a typo.

commit 0499c545430edadfc864d5b1ac8096952d6d0f1a
Merge: c3dcb7c bcaca3a
Author: Daniel Arndt <daniel.arndt@iwr.uni-heidelberg.de>
Date:   Fri Sep 23 17:23:00 2016 +0200

    Merge pull request #3152 from bangerth/doc-update-77
    
    Use latex formulas to describe tensor invariants.

commit c3dcb7c6b10b4ee751379bae16d0becd57510657
Merge: ea0d225 f1e7af8
Author: Denis Davydov <davydden@gmail.com>
Date:   Fri Sep 23 06:08:47 2016 +0200

    Merge pull request #3150 from masterleinad/transfer_constraints
    
    Eliminate unused ConstraintMatrix objects in MGTransfer
```

**Comment by bangerth** [comment]
We merge hundreds of pull requests per release, and most of those have no place appearing in the changelog because they are of no interest to the user. The three you cite above are good examples: relevant patches, no value to a human reader. Who is going to go through hundreds of these sorts of messages, format them for human readers, and collate them into anything that's useful to our users?

Most projects struggle to find reviewers. I'm totally against making reviewers' life even more onerous. That burden should be on the author of the patch.

**Comment by tamiko** [comment]
> Most projects struggle to find reviewers. I'm totally against making
> reviewers' life even more onerous. That burden should be on the author
> of the patch.

?

I don't understand. With my suggested approach no pull request without a
meaningful changelog entry would get merged. (We have quite a number of
"oops we forgot the changes.h entry pull requests...)

The burden would be at the author of the patch.

**Comment by tamiko** [comment]
> We merge hundreds of pull requests per release, and most of those have no place appearing in the changelog because they are of no interest to the user.

Yes. And selecting meaningful merges as candidates for the changelog entries would be as simple as invoking `git log` with either a filter for a keyword (like `Authors:`), or with a filter for showing only merge messages that are longer than one line.

**Comment by tamiko** [comment]
But anyway. I let you guys decide what's best and will not discuss any further. For better readability, I remove my last comments.

**Comment by jppelteret** [comment]
Although there's no perfect option here, I like the suggestion of a set of separate incremental change files the most. Although they're yet another thing to maintain, at least they're maintainable in the sense that their author can easily reword them any time up will they're added to `changes.h`.

Regarding categorisation, I'm on board with the idea of categorising everything that isn't an incompatibility just before release. The only point I'd be concerned about is that the new, exciting features aren't immediately highlighted to the users of the development version.

**Comment by kronbichler** [comment]
> The only point I'd be concerned about is that the new, exciting features aren't immediately highlighted to the users of the development version. 

Are they now? For the really big changes like new tutorials we still have the "News" page.

**Comment by bangerth** [comment]
Yes: https://www.dealii.org/developer/doxygen/deal.II/changes_after_8_4_2.html

(I do think that a fair number of entries in the "general improvements" section are misplaced and should go to "specific changes", as pointed out further up in the thread somewhere.)

**Comment by davydden** [comment]
maybe then we should additionally do preliminary categorization into `minor`/`major` (or whatever names) and run scripts to generate `changes.h` roughly every month so that users see what are the changes and whether or not they need to use `develop`. 

Before release, things will be shuffled within those two categories.

**Comment by bangerth** [comment]
My idea was to run the script that builds `changes.h` every time the documentation is generated (which is every hour or so for the website).

**Comment by jppelteret** [comment]
@kronbichler Fair point, evidently they're not.

**Comment by kronbichler** [comment]
@bangerth @jppelteret That is my point: We technically have the distinction major/minor through general/specific, but it is not really good. I like everyone else tend to over-estimate the "coolness" of my own developments with respect to those two lists, even though I try to be conservative.

**Comment by tjhei** [comment]
> I like everyone else tend to over-estimate the "coolness" of my own developments with respect to those two lists, even though I try to be conservative.

I just stick it randomly into one of the two categories. ;-)
